### ISSUE TITLE
esbuild does not fail with undefined functions

### ISSUE THREAD

**Issue by dewang** [evanw/esbuild#3719]
I would expect esbuild to fail if the code to be built has calls to functions that do not exist. i.e. they are either not imported, or not defined in the code being bundled. However, it seems to simply ignore all unknown functions and produce output successfuly,

Simplest possible example shown here -

https://esbuild.github.io/try/#YgAwLjIwLjIAaW5kZXgudHMgLS1idW5kbGUAAGluZGV4LnRzAGV4cG9ydCBjb25zdCBmb28gPSAoKSA9PiB7CiAgYmFyKCk7Cn0

**Comment by hyrious** [comment]
This is because esbuild doesn't have a proper list about callable function names. For example, you can call `fetch()` at any where as long as the environment supports it (in the browser, or Node.js 18+). The `undefined` checking should be done in the TypeScript compiler since it has `lib.*.ts` and `@types/node`.

**Comment by dewang** [comment]
Yes but esbuild bundling manages the conversion from Typescript to JS so I would expect it to handle that. I do not have a separate step where we call the Typescript compiler.

**Comment by lazarljubenovic** [comment]
You need TypeScript for type-checking anyway. https://esbuild.github.io/try/#YgAwLjIwLjIAaW5kZXgudHMgLS1idW5kbGUAAGluZGV4LnRzAGV4cG9ydCBjb25zdCBhOiBudW1iZXIgPSB0cnVl

**Comment by dewang** [comment]
Perhaps I'm missing something but I don't see where type checking is happening in the link that you shared. It will still compile anything you type in the .ts file into .js.

**Comment by evanw** [comment]
I'm closing this issue because type checking is out of esbuild's scope. You need to run the TypeScript type checker to check the types in your TypeScript code. This is already called out in esbuild's documentation: https://esbuild.github.io/content-types/#typescript

> esbuild has built-in support for parsing TypeScript syntax and discarding the type annotations. However, esbuild _does not_ do any type checking so you will still need to run `tsc -noEmit` in parallel with esbuild to check types. This is not something esbuild does itself.

The main reason for this is because the TypeScript compiler is both incredibly complex and ever-changing. Replicating the TypeScript compiler along with all of its implementation quirks and keeping up with the ongoing changes made by a whole team of paid Microsoft employees would be a significant undertaking, and is also a mostly independent problem from stripping TypeScript types and bundling code. Something like that is best done as a separate project. Parsing TypeScript and stripping types is a simple task by comparison, so I feel ok including that in esbuild's scope.
